staging/src/k8s.io/component-base/logs/datapol/datapol.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+// Package datapol contains functions to determine if objects contain sensitive
+// data to e.g. make decisions on whether to log them or not.
+package datapol
+
+import (
+	"reflect"
+	"strings"
+)
+
+// Verify returns a list of the datatypes embeded in the argument that can be
+// considered sensitive w.r.t. to logging
+func Verify(value interface{}) []string {
+	t := reflect.ValueOf(value)
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+	return datatypes(t)
+}
+
+func datatypes(v reflect.Value) []string {
+	switch v.Kind() {
+	case reflect.Slice, reflect.Array:
+		for i := 0; i < v.Len(); i++ {
+			if types := datatypes(v.Index(i)); len(types) > 0 {
+				return types
+			}
+		}
+	case reflect.Map:
+		mapIter := v.MapRange()
+		for mapIter.Next() {
+			k := mapIter.Key()
+			v := mapIter.Value()
+			if types := datatypes(k); len(types) > 0 {
+				return types
+			}
+			if types := datatypes(v); len(types) > 0 {
+				return types
+			}
+		}
+	case reflect.Struct:
+		if types := byType(v.Type()); len(types) > 0 && !v.IsZero() {
```

Comment:
Was checking if element value is `IsZero` proposed in original KEP? 

Would like to get @PurelyApplied opinion

Reply:
From the KEP: 
> Values of individual fields of primitive types like string will not be checked other than checking if a given field is not empty.

My reading of that is that this implies the IsZero check

Reply:
Thanks for pointing this out. This surprised me as this introduces additional overhead to check every element instead of type of first element and doesn't make sense if we want to implement caching based on the type.

@44past4 WDYT?

Reply:
As vast majority of the types being logged should not have any datapol tags therefore I believe that caching based on type should still work pretty good.

Reply:
The vast majority of logging arguments in Kubernetes are either a primitive types or simple structs. Therefore I strongly believe that caching by type we will be able to cover something like 99% of actual log arguments and therefore significantly improve performance. But as I do not have a hard data on this I agree that we should leave this to further performance optimization phase of the implementation.

Reply:
Okay, lets leave the caching/optimization discussion for now, however I feel like the IsZero needs to stay, so resolving this conversation.